tests/utils/test_python_virtualenv.py

```diff
@@ -116,25 +116,20 @@ def test_should_create_virtualenv_with_extra_packages(self, mock_execute_in_subp
         mock_execute_in_subprocess.assert_called_with(["/VENV/bin/pip", "install", "apache-beam[gcp]"])
 
     def test_remove_task_decorator(self):
-        py_source = "@task.virtualenv(use_dill=True)\ndef f():\nimport funcsigs"
+        py_source = "@task.virtualenv(use_dill=True)\ndef f(): ...\nimport funcsigs"
         res = remove_task_decorator(python_source=py_source, task_decorator_name="@task.virtualenv")
-        assert res == "def f():\nimport funcsigs"
+        assert res == "def f():\n    ...\nimport funcsigs"
 
     def test_remove_decorator_no_parens(self):
-        py_source = "@task.virtualenv\ndef f():\nimport funcsigs"
+        py_source = "@task.virtualenv\ndef f(): ...\nimport funcsigs"
         res = remove_task_decorator(python_source=py_source, task_decorator_name="@task.virtualenv")
-        assert res == "def f():\nimport funcsigs"
-
-    def test_remove_decorator_including_comment(self):
-        py_source = "@task.virtualenv\ndef f():\n# @task.virtualenv\nimport funcsigs"
-        res = remove_task_decorator(python_source=py_source, task_decorator_name="@task.virtualenv")
-        assert res == "def f():\n# @task.virtualenv\nimport funcsigs"
```

Comment:
I'm not sure if it is fine to remove the comments after the `ast.unparse`. If it is not an allowed behavior, we might not be able to adopt ast as a solution, since it would not preserve comments.

Reply:
I think it would be better to use libcst https://pypi.org/project/libcst/ - which we alredy consider to use for our "upgrade check" tooling -> https://github.com/apache/airflow/issues/41641#issuecomment-2439879558

this way we could preserve comments, line numbers etc. I think the important thing we might want to keep here is debuggability and particularly line number references, and with AST we will likely loose it - not only the comments.

Reply:
cc: @jscheffl @uranusjr @ashb @kaxil @tirkarthi ?

Reply:
I have no expertise in this matter so I can not judge this.

In general I'd favor is we refactor to ensure pytests are staying the same. I am also not sure why the line feed was removed in the other tests. Is this a compatibility limitation or just a cleanup?

Reply:
It's the difference of AST vs. CST. 

* (AST) Abstract Syntax Tree - > is the tree of syntax of Python code that contains only the "meaningful" tokens -  that represent a "runtime" python bytecode (with stripped out comments) - https://en.wikipedia.org/wiki/Abstract_syntax_tree


* (CST) Concrete Syntax Tree -> is the tree of syntax of Python code that represent the `source code` - not Python runtime bytecode (that includes EOL characters and comments - but also all other non-runtime tokens - https://en.wikipedia.org/wiki/Parse_tree

Reply:
(that's a bit simplified definition of course - but I think it describes the difference well).

Reply:
If we're doing code modification instead of checking only, we probably should go with libcst I think 🤔

Reply:
I happened to share this topic at PyCon Taiwan this year. If anyone is interested, this was the slide I used. The content around page 67 is the most relevant.

https://speakerdeck.com/leew/unleash-the-chaos-developing-a-linter-for-un-pythonic-code?slide=67

Reply:
Yeah, I think it's better to leave dag authors' code information as much as possible in the rendered `script.py` in virtualenv operator, ideally the unit tests should not be changed, let me try LibCST instead of AST here.

Reply:
If you need any example, I used it to check default_deferrable here https://github.com/apache/airflow/blob/main/scripts/ci/pre_commit/check_deferrable_default.py